akka-typed/src/main/scala/akka/typed/Behavior.scala

```diff
@@ -132,12 +156,24 @@ object Behavior {
    * behavior) this method computes the next behavior, suitable for passing a
    * message or signal.
    */
-  def canonicalize[T](behavior: Behavior[T], current: Behavior[T]): Behavior[T] =
+  def canonicalize[T](behavior: Behavior[T], current: Behavior[T], ctx: ActorContext[T]): Behavior[T] =
+    behavior match {
+      case SameBehavior                  ⇒ current
+      case UnhandledBehavior             ⇒ current
+      case deferred: DeferredBehavior[T] ⇒ canonicalize(undefer(deferred, ctx), deferred, ctx)
+      case other                         ⇒ other
+    }
+
+  def undefer[T](deferredBehavior: DeferredBehavior[T], ctx: ActorContext[T]): Behavior[T] = {
+    val behavior = deferredBehavior(ctx)
+
     behavior match {
-      case `sameBehavior`      ⇒ current
-      case `unhandledBehavior` ⇒ current
-      case other               ⇒ other
+      case innerDeffered: DeferredBehavior[T] ⇒ undefer(innerDeffered, ctx)
+      case ext: ExtensibleBehavior[T] ⇒ ext
+      case StoppedBehavior | EmptyBehavior | UnhandledBehavior | SameBehavior ⇒ behavior
+      case x ⇒ throw new IllegalStateException(s"deferred cannot return $x as behavior")
```

Comment:
When I think of it it's not correct either, because `Same` for example would be ok when used inside some other behavior, just not outermost, which `Behavior.preStart` should already guard against